### ISSUE TITLE
`[BFTask taskForCompletionOfAllTasksWithResults:tasks]` (BUT IN SERIES)

### ISSUE THREAD

**Issue by tettoffensive** [BoltsFramework/Bolts-ObjC#272]
I have an undefined number of tasks that I would normally run in parallel. but they involve AVFoundation encoding which errors out when the tasks run in parallel. So I'd ideally like a similar method that just chained them together with the final `task.result` being the array of tasks (just as `taskForCompletionOfAllTasksWithResults` has.

Right now I'm just looping through the array myself and chaining them together. But in this case I only have the result from the task at the end of the chain.

```
BFTask *taskChain = nil;
    for (BFTask *task in tasks) {
        if (!taskChain) {
            taskChain = task;
        } else {
            taskChain = [taskChain continueWithSuccessBlock:^id _Nullable(BFTask * _Nonnull task) {
                return task;
            }];
        }
    }
    
```

**Comment by nlutsenko** [comment]
Hey @tettoffensive, so the thing here is that if you have an array of `tasks`, that would mean that they are already either running or are scheduled to be started.

This said, we could do something with execution blocks, where you would need to pass an array of them, or even better - something that we built internally (still in open source, but not public API) for Parse SDK - `TaskQueue`.

I would suggest taking a look at the queue implementation [here](https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/blob/master/Parse/Internal/PFAsyncTaskQueue.m).
Does it solve your problem? If yes - we can introduce it to core Bolts set of APIs.

**Comment by tettoffensive** [comment]
@nlutsenko This sounds like it could be quite useful. What I ended up doing for now is writing a recursive function that would just continue to append BFTasks until an array of source data is empty.

**Comment by caoer** [comment]
@nlutsenko @tettoffensive I got a naive idea which basically has a generator which created the task so that it is executing serialized. It does have a problem on which queue the `self.generator()` is called. but the interface looks very similar to taskForCompletionOfAllTasks. what do you think about it? 

```
@interface BFTaskColdGenerator()
@property (nonatomic, copy) BFTask* (^generator)(void);
@end

@implementation BFTaskColdGenerator
- (instancetype)initWithGeneratorBlock:(BFTask *(^)(void))generator {
    self = [super init];
    if (self) {
        self.generator = generator;
    }
    return self;
}

- (BFTask *)task {
    return self.generator();
}
@end


@implementation BFTask (ColdGenerator)

+ (NSArray<BFTask *> *)tasksForCompletionOfAllTasksSerialWithTaskGenerators:(NSArray<BFTaskColdGenerator *> *)taskGenerators {
    
    __block NSMutableArray<BFTaskColdGenerator *> *mutableTaskGenerator = [taskGenerators mutableCopy];
    
    __block NSMutableArray *results = [NSMutableArray array];
    
    while (mutableTaskGenerator.count) {
        BFTaskColdGenerator *tcg = mutableTaskGenerator.firstObject;
        BFTask *task = tcg.task;
        [task waitUntilFinished];
        [task continueWithBlock:^id _Nullable(BFTask * _Nonnull t) {
            [mutableTaskGenerator removeObjectAtIndex:0];
            [results addObject:t];
            return nil;
        }];
    }
    
    return results;
}

@end
```

Tests:

```
- (void)testExample {
    NSUInteger count = 50;
    NSMutableArray *tasks = [NSMutableArray arrayWithCapacity:count];
    
    for (NSUInteger i = 0; i < count; i++) {
        int r = arc4random() % 100;
        [tasks addObject:@[@(i), @(r)]];
    }
    
    NSMutableArray *taskGenerator = [NSMutableArray array];
    
    for (NSArray *taskValue in tasks) {
        [taskGenerator addObject:[[BFTaskColdGenerator alloc] initWithGeneratorBlock:^BFTask * _Nonnull{
            return [[BFTask taskWithDelay:[taskValue[1] intValue]] continueWithBlock:^id _Nullable(BFTask * _Nonnull t) {
                NSLog(@"executing task: %@, delay: %@", taskValue[0], taskValue[1]);
                return t;
            }];
        }]];
    }
    
    [BFTask tasksForCompletionOfAllTasksSerialWithTaskGenerators:taskGenerator];
}
```

**Comment by caoer** [comment]
using PFAsyncTaskQueue implementation much cleaner: 
```
+ (BFTask *)tasksForCompletionOfAllTasksSerialWithTaskGenerators:(NSArray<BFTaskColdGenerator *> *)taskGenerators {
    PFAsyncTaskQueue *queue = [[PFAsyncTaskQueue alloc] init];
    
    __block BFTask *task;
    for (BFTaskColdGenerator *generator in taskGenerators) {
        task = [queue enqueue:^id _Nullable(BFTask * _Nonnull t) {
            return generator.task;
        }];
    }
    
    return task;
}
```

**Comment by swordfishyou** [comment]
I also faced the same kind of situations where I have to run some number of tasks in series. Task queue was a perfect solution, but it'd be nice to have this functionality by default, as mentioned above.
